### DISCUSSION TITLE
Config Redirects: Negative `has` condition

### DISCUSSION THREAD

**Discussion by hpohlmeyer** [vercel/next.js#28186] [Ideas]
### Describe the feature you'd like to request

Next.js Redirects in `next.config.js` allow for an additional condition using the `has` property. 

```tsx
async redirects() {
    return [
      {
        source: '/specific/:path*',
        has: [
          {
            type: 'cookie',
            key: 'authorized',
            value: 'true',
          },
        ],
        permanent: false,
        destination: '/another/:path*',
      },
    ]
}
```

A lot of times you want to check for the absence of a header, cookie or query instead. This would be a nice addition here.

### Describe the solution you'd like

I would like to see a similar approach as the `has` property, but with a `hasNot` property. The array content should use the same format as the `has` prop, but check for the absence of the defined entries. e.g.

```tsx
async redirects() {
    return [
      {
        source: '/:path*',
        hasNot: [
          {
            type: 'cookie',
            key: 'authorized',
            value: 'true',
          },
        ],
        permanent: false,
        destination: '/login',
      },
    ]
}
```

should only redirect to the login page, if there is no `authorized` cookie withe a value of `true`. It should also be possible to use wildcards for the `value` property.

### Describe alternatives you've considered

__1. Redirect in `getServerSideProps`__  
If you do SSR, you could check the values in `getServerSideProps`, where you can access the request. If you need global redirects, this would mean you need to copy the logic for it into every page, making it a big maintainance burden. But it is even worse for static pages, because the logic would not work on those pages.

__2. Client-side checks__  
Query parameters could be checked on the client side, as well as non-http-only-cookies. Afterwards you could use next router to redirect, but that results in extra round trips, when a server-redirect would be preferable.

__3. Pass requests through an API-route__  
You could also pass all your requests through an API route. In that API route you have access to the headers and http-only cookies and you could set another cookie thats used as a flag for a non-existent cookie. Afterwards you redirect to your original requested page.
You could now check for the flag-cookie in your `next.config.js`-redirects.

**Comment by chemicalkosek** [comment]
https://github.com/vercel/next.js/discussions/24650
